hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/updatetokencustomfees/UpdateTokenCustomFeesDecoder.java

```diff
@@ -162,52 +157,60 @@ private Stream<CustomFee> decodeFractionalFees(
     }
 
     private Stream<CustomFee> decodeRoyaltyFees(
-            @NonNull final Tuple[] royaltyFees, @NonNull final AddressIdConverter addressIdConverter) {
+            @NonNull final Tuple[] royaltyFees, @NonNull final HtsCallAttempt attempt) {
         return Arrays.stream(royaltyFees).map(fee -> CustomFee.newBuilder()
                 .royaltyFee(RoyaltyFee.newBuilder()
                         .exchangeValueFraction(Fraction.newBuilder()
                                 .numerator(fee.get(ROYALTY_FEE_NUMERATOR))
                                 .denominator(fee.get(ROYALTY_FEE_DENOMINATOR))
                                 .build())
-                        .fallbackFee(getFallbackFee(fee))
+                        .fallbackFee(getFallbackFee(attempt, fee))
                         .build())
-                .feeCollectorAccountId(addressIdConverter.convert(fee.get(ROYALTY_FEE_FEE_COLLECTOR)))
+                .feeCollectorAccountId(attempt.addressIdConverter().convert(fee.get(ROYALTY_FEE_FEE_COLLECTOR)))
                 .build());
     }
 
     // In the solidity structure for Fixed Fees we have the property `bool useCurrentTokenForPayment` that is not
     // present in the protobuf version.
     // In order to determine the denominating token we need to check if the property is present.
-    private TokenID determineDenominatingToken(@NonNull final Tuple fee, @NonNull final TokenID tokenId) {
+    private TokenID determineDenominatingToken(
+            @NonNull final HtsCallAttempt attempt, @NonNull final Tuple fee, @NonNull final TokenID tokenId) {
         final boolean useCurrentToken = fee.get(FIXED_FEE_USE_CURRENT_TOKEN_FOR_PAYMENT);
         return useCurrentToken
                 ? tokenId
-                : getDenominationTokenIdOrNull(fee.get(FIXED_FEE_TOKEN_ID), fee.get(FIXED_FEE_USE_HBARS_FOR_PAYMENT));
+                : getDenominationTokenIdOrNull(
+                        attempt, fee.get(FIXED_FEE_TOKEN_ID), fee.get(FIXED_FEE_USE_HBARS_FOR_PAYMENT));
     }
 
     // In the solidity structure for FixedFees we have the property `bool useHbarsForPayment` that is not present
     // in the protobuf version where we set the denominating token to null if the fee is in hbars and/or we provide zero
     // address.
     private @Nullable TokenID getDenominationTokenIdOrNull(
-            @NonNull final Address address, final boolean useHbarsForPayment) {
-        final var tokenId = ConversionUtils.asTokenId(address);
-        return useHbarsForPayment || tokenId.equals(TokenID.DEFAULT) ? null : tokenId;
+            @NonNull final HtsCallAttempt attempt, @NonNull final Address address, final boolean useHbarsForPayment) {
+        final var tokenId = ConversionUtils.asTokenId(attempt.nativeOperations().entityIdFactory(), address);
+        return useHbarsForPayment
+                        || tokenId.equals(
+                                attempt.nativeOperations().entityIdFactory().newTokenId(0))
```

Comment:
Seems like unnecessary work to convert `address` to a fully-qualified tokenId, convert `0` to a token ID, and then compare them. What about something similar to this?

```suggestion
        final var tokenNum = ConversionUtils.extractTokenNum(address);
        return useHbarsForPayment || tokenNum == 0
```